### PR TITLE
iCubGenova04,iCubGenova02: Upgrade the full robot configuration to be compatible with the new "flexible motor input" motor control interface

### PR DESCRIPTION
**Pull request by nunoguedelha**
### Changes Overview
These changes are the first step in the integration and test of the low level current control through the `ICurrentControl` interface, in a seamless way with respect to the other control interfaces (Position, velocity and torque). Refer to https://github.com/dic-iit/element_torque-control-via-current/issues/58 for more details.
### Quick insight on the impact of the new `ICurrentControl` interface on the other controllers
Previous to the implementation of the "flexible motor input"(*) by @ale-git, **each controller** (Position, Position Direct, Torque, Velocity) **running in the EMS board @1KHz** was generating the same type of output, a PWM setpoint, from the desired respective position, torque or velocity and respective feed-forward and PID parameters. That output was then fed to the **motor driver running in the 2FOC @20KHz**, that would modulate the PWM on each motor phase according to the rotor position.

After the implementation of the "flexible motor input", **each controller can output a desired PWM, current or velocity** that is fed to the motor driver (2FOC), where it is converted to the ultimate PWM setpoint. In the case of a desired current input, that driver closes the loop on the measured current (@20KHz).

In the default configuration of the new interface implementation, the controllers output **was set to a desired current**. we are changing this output to a desired PWM, to match the same behavior as before the "flexible motor input" implementation.

The change is to be done through the device configuration files in `robots-configuration/<robot-name>/hardware/motorcontrol`.

(*) The list of issues will be provided soon.

### Implementation Details
The changes are based on the initial current control configuration implemented by @ale-git for the "flexible motor input" feature: https://github.com/robotology/robots-configuration/commit/421b27823cf83554aa69b3938637b280edab2569 and https://github.com/robotology/robots-configuration/commit/430b61f510d0f5b543edbd1fe4d2aaf784c2905a (branch https://github.com/robotology/robots-configuration/tree/icub3_experimental_velcontrol)

They include:
1. a dummy current control configuration for all the robot joints, compatible with the new controller interfaces ("flexible motor input") implemented in `YARP` and `icub-main`, such that the robot interface can start,
2. the initial current control configuration for the left leg implemented in https://github.com/robotology/robots-configuration/commit/421b27823cf83554aa69b3938637b280edab2569
3. the presently used configuration of the other control modes: position, torque.

### Remaining Actions
- [x] push an update with the Single Motor Setup files moved out of iCubGenova04
- [x] ~Add to the test setup config the files used on the MOOG V3~ (will do in a later PR, not a priority now).
- [x] include also @randaz81 in the reviewers.
- [x] push a config for iCubGenova02 compatible with the `icub-main` changes to be merged to `devel` (https://github.com/robotology/icub-main/pull/585).
- [x] same for iCubV3 and R1. => TO BE DONE BY @ale-git 

This PR is blocked by https://github.com/robotology/icub-main/pull/585.

### CODE REVIEW

**Comment by nunoguedelha** [comment]
Ciao @S-Dafarra , for your information, I have checked that the last commit of this PR is identical to the respective commit already tested on the robot (abc0fdc44bffc9d754edd474ecd34be70e1bd030) prior the rebase.
These changes are specific to the "flexible motor input" feature (new motor current control) and do not impact the robot dynamic behavior. But the parsing of the position and torque controller parameters is slightly impacted which results in dependencies in `icub-main`, `icub-firmware-shared` and `icub-firmware-build` currently built and flashed (EMS and 2FOC firmware) on the robot head.

**Comment by S-Dafarra** [comment]
Hi @nunoguedelha! Yes, my main concern would be about the testing with the robot. Since it changes the configuration files for each control mode, I would like to be sure that each demo (at least the ones we are showing more frequently) is still working as expected, before merging. 

Also, just to be sure, are ``icub-main``, ``icub-firmware-shared`` and ``icub-firmware-build`` all in ``devel`` or in ``master``?

I was also a bit concerned about the fact that we are adding new configuration files in ``hardware/motorControl``. I am afraid that this could generate confusion on which file is actually loaded in normal conditions. Maybe it could be reasonable to create a folder called ``reducedConfigurations`` (or whatever clearer name) and move those files to ``reducedConfigurations/hardware/motorControl``. Would this be feasible?

**Comment by nunoguedelha** [comment]
> Hi @nunoguedelha! Yes, my main concern would be about the testing with the robot.

Yes, the files pointed by this PR and included in the `robots-configuration/iCubGenova04/` are **identical** to the ones pointed by the commit `https://github.com/robotology/robots-configuration/commit/abc0fdc44bffc9d754edd474ecd34be70e1bd030` on which the main demos have already been tested.

> Also, just to be sure, are `icub-main`, `icub-firmware-shared` and `icub-firmware-build` all in `devel` or in `master`?

Unfortunately only the `icub-firmware-build` changes are in `devel` (refer to https://github.com/dic-iit/element_torque-control-via-current/issues/57). I totally agree that the proper course of action is to not merge the `robots-configuration` changes without the dependencies being properly pushed to the `devel` branch on the respective repos. The problem is that the `icub-main` changes cannot be merged to `devel` without propagating the respective `robots-configuration` changes to all the **ethernet robots** using the `devel` branch (as per @ale-git this boils down to iCubGenova02 and iCubGenova04).

As a side note, as previously discussed with @DanielePucci, the idea of the early integration of `robots-configuration` changes in the Green iCub code was to be able to switch easily between demo tests and "flexible motor input" feature tests, having the boards flashed once and for all.

I'm currently discussing with the @ale-git to merge all the required changes into `devel` earlier than planned, ideally this week or early next week.



> I was also a bit concerned about the fact that we are adding new configuration files in `hardware/motorControl`

True. Initially this was iCubGenova04 specific, but as you say it makes more sense anyway to create a "new" robot folder at the same level as `iCubGenova04`. I have to check with the other repo maintainers that this doesn't create unexpected problems or requires some tools updates (typically the automatic generation of configuration files @randaz81 worked on some time ago).

**Comment by S-Dafarra** [comment]
> Yes, the files pointed by this PR and included in the robots-configuration/iCubGenova04/ are identical to the ones pointed by the commit https://github.com/robotology/robots-configuration/commit/abc0fdc44bffc9d754edd474ecd34be70e1bd030 on which the main demos have already been tested.

Ok great!

>  The problem is that the icub-main changes cannot be merged to devel without propagating the respective robots-configuration changes to all the ethernet robots using the devel branch

Ah this is a chicken-egg problem :sweat_smile:. Are they back compatible, i.e. can the robot run if we move to ``icub-main`` to ``devel``?

> True. Initially this was iCubGenova04 specific, but as you say it makes more sense anyway to create a "new" robot folder at the same level as iCubGenova04.

I was actually meaning inside the ``iCubGenova04`` folder to avoid confusion with the normal configuration files. I don't know whether this applies to all robots.

**Comment by nunoguedelha** [comment]
> I would prefer not to use a particular branch of `icub-main` for too long.

Absolutely. We will have a meeting on Thursday with @randaz81 (to be confirmed) in order to discuss the merge of the `icub-main` and `icub-firmware-shared` changes to `devel`. We will then make a choice accordingly.

> I was actually meaning inside the `iCubGenova04` folder ...

Sorry I wasn't very clear on this.. 
- The new configuration files `knee_...` actually reside in multiple folders `hardware/motorControl`, `hardware/electronics`, ... in order to replicate the configuration of a single or a set of joints, only in this case, a single joint is set with a MOOG motor and gearbox (harmonic drive).
- The folder structure matches the same as the one used for any other robot.
- This configuration allows to run a robot interface on a single motor setup that can be used to test a MOOG of iCub2.5 as well as a MOOG of iCubV3, or even a Koolmorgen (old iCub2), through small configuration changes.

That's why I was thinking at moving the files to a standalone folder `robotology/robots-configuration/<some_setup_name>/`. (Don't know yet which name to pick).

**Comment by S-Dafarra** [comment]
> Absolutely. We will have a meeting on Thursday with @randaz81 (to be confirmed) in order to discuss the merge of the icub-main and icub-firmware-shared changes to devel. We will then make a choice accordingly.

Ok great! Let me know what you decide then.

> That's why I was thinking at moving the files to a standalone folder robotology/robots-configuration/<some_setup_name>/. (Don't know yet which name to pick).

Is there a specific reason why you prefer to keep the ``<some_setup_name>`` folder as a standalone one and not in ``robots-configuration/iCubGenova04/<reduced_setups_folder>/<some_setup_name>``? I think it may make sense to keep it inside the ``iCubGenova04`` folder as those configuration files have been tested and used only on the greeny, but it is just a suggestion.

**Comment by nunoguedelha** [comment]
> those configuration files have been tested and used only on the greeny

Mostly for three reasons:
- the committed files are the ones tested on a MOOG V2.5, which is the exact same motor as the ones used on iCubGenova02. The configuration was tuned for a standalone motor (no hw nor sw limits, different offsets), unmounted from some iCub V2.5. It wasn't "tested" on the greeny, it was tested on a bench with a single motor.
- actually the same config with few changed params (`resolution` sign, `HasHallSensor`, `HasRotorEncoderIndex`) was tested on a MOOG V3. I should actually commit new files with those configs, to make it clearer. Example:
```
hardware/mechanicals/knee-eb10-j0-mec-V25.xml
hardware/mechanicals/knee-eb10-j0-mec-V30.xml etc
```
- we keep the same folder structure as other robot setups
- we don't pollute the iCubGenova04 folder with files that are not actually run on the robot but on a standalone setup. Although the setup was initially used to test a problem detected on iCubGenova04, that problem was also spotted on iCubV3
- If hypothetically this setup is also used for testing the friction and we decide to add an accelerometer to the shaft for instance, we can add a folder `inertials` to the configuration.

What you should picture is: this setup is not specific to iCubGenova04.

**Comment by S-Dafarra** [comment]
> What you should picture is: this setup is not specific to iCubGenova04.

Much clearer now, thanks! I agree in moving them out of the ``iCubGenova04`` folder then!

**Comment by nunoguedelha** [comment]
I'll push an update with the moved files and include also @valegagge in the reviewers.

Probably I'll have to push a config for iCubGenova02 in case the `icub-main` changes will be pushed soon enough to `devel`.

**(added these actions in the description)**

**Comment by nunoguedelha** [comment]
ciao @ale-git , could you please do the pull request for R1, such that @randaz81 can run the tests on R1 with the official branch asap? thx.

**Comment by nunoguedelha** [comment]
> ciao @ale-git , could you please do the pull request for R1, such that @randaz81 can run the tests on R1 with the official branch asap? thx.

PR: https://github.com/robotology/cer/pull/105

**Comment by nunoguedelha** [comment]
After the latest minor fixes (https://github.com/robotology/robots-configuration/pull/100/commits/558e4203d871f64223abbb30e4a250f0f6caadb7) on `iCubGenova02`, pushed and tested, the PR is ready for the final review.

**Comment by nunoguedelha** [comment]
@S-Dafarra @ale-git @randaz81 could you please review the latest changes.

The plan is to merge the pull request tomorrow along with the dependencies.

Thx!

File: iCubGenova04/icub_right_leg.xml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        
+    <devices>
+    <!-- HEAD -->
+    <!--<xi:include href="./hardware/motorControl/head-eb20-j0_1-mc.xml" />  
+    <xi:include href="./hardware/motorControl/head-eb21-j2_5-mc.xml" /> 
+    <xi:include href="./wrappers/motorControl/head-mc_wrapper.xml" /> -->
+ 
+    <!-- TORSO --> 
+    <xi:include href="hardware/motorControl/torso-eb5-j0_2-mc.xml" />
+    <xi:include href="wrappers/motorControl/torso-mc_wrapper.xml" />
```
Comment (S-Dafarra) [inline iCubGenova04/icub_right_leg.xml]: This file is called ``right_leg`` but it starts the whole robot. Maybe a different name?

Comment (nunoguedelha) [reply iCubGenova04/icub_right_leg.xml]: Actually that file is not required anymore. The robot can start with any of the other scripts. So I'll just remove it. I'll wait for the other comments to push the update.

Comment (nunoguedelha) [reply iCubGenova04/icub_right_leg.xml]: sorry @S-Dafarra the PR has been merged without the change on the cleanup removal of the `icub_right_leg.xml` script. I will do it in my next PR this week.

**Pull request merged**